File: built-in plugins dsp/mb-time-pitchtime.cpp
/**
 * MB Pitch-Time
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Independent pitch and time manipulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TIME_PITCHTIME_H
#define MB_TIME_PITCHTIME_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTimePitchtime : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-time-pitchtime";
    static constexpr const char* PLUGIN_NAME    = "MB Pitch-Time";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float pitch = 0f;  // range [-24, 24]
    float timeRatio = 1f;  // range [0.25, 4]
    float formant = 0f;  // range [-12, 12]
    float mix = 1f;  // range [0, 1]
    };

    MbTimePitchtime() = default;
    ~MbTimePitchtime() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.pitch = std::clamp(params.pitch, -24f, 24f);
        params.timeRatio = std::clamp(params.timeRatio, 0.25f, 4f);
        params.formant = std::clamp(params.formant, -12f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Pitch-Time
        return input;
    }
};

#endif // MB_TIME_PITCHTIME_H

File: built-in plugins dsp/mb-time-pitchtime.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbTimePitchtimePlugin: PluginDefinition = {
  id: "mb-time-pitchtime",
  slug: "mb-time-pitchtime",
  name: "MB Pitch-Time",
  category: "effect",
  type: "delay" as any,
  version: "1.0.0",
  description: "Independent pitch and time manipulation",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "pitch",
      name: "Pitch",
      type: "float",
      defaultValue: 0,
      minValue: -24,
      maxValue: 24,
      automatable: true,
    },
    {
      id: "timeRatio",
      name: "Time Ratio",
      type: "float",
      defaultValue: 1,
      minValue: 0.25,
      maxValue: 4,
      automatable: true,
    },
    {
      id: "formant",
      name: "Formant",
      type: "float",
      defaultValue: 0,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { pitch: 0, timeRatio: 1, formant: 0, mix: 1 },
};

export default MbTimePitchtimePlugin;
